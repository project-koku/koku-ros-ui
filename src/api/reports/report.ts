import type { PagedMetaData, PagedResponse } from 'api/api';

export interface ReportValue {
  count?: number;
  count_units?: string;
  units?: string;
  unused?: number;
  unused_percent?: number;
  value?: number;
}

export interface ReportItemValue {
  distributed?: ReportValue;
  markup?: ReportValue;
  platform_distributed?: ReportValue;
  raw?: ReportValue;
  total?: ReportValue;
  usage?: ReportValue;
  worker_unallocated_distributed?: ReportValue;
}

export interface ReportItem {
  classification?: string; // Platform.
  cost?: ReportItemValue;
  date?: string;
  default_project?: string; // 'True' or 'False'
  delta_percent?: number;
  delta_value?: number;
  infrastructure?: ReportItemValue;
  source_uuid?: string;
  supplementary?: ReportItemValue;
}

export interface ReportAwsItem extends ReportItem {
  account?: string;
  account_alias?: string;
  region?: string;
  service?: string;
}

export interface ReportGcpItem extends ReportItem {
  account?: string;
  project?: string;
  region?: string;
  service?: string;
}

export interface ReportAzureItem extends ReportItem {
  resource_location?: string; // 'region'
  service_name?: string; // 'service'
  subscription_guid?: string; // 'account'
}

export interface ReportOcpItem extends ReportItem {
  capacity: ReportValue;
  cluster?: string;
  clusters?: string[];
  limit: ReportValue;
  node?: string;
  persistentvolumeclaim?: string;
  project?: string;
  request: ReportValue;
  usage: ReportValue;
}

export interface ReportOrgItem extends ReportItem {
  alias?: string;
  id?: string;
}

// Additional props for group_by[org_unit_id]
export interface ReportOrgData {
  id?: string;
  type?: string; // 'account' or 'organizational_unit'
}

// Additional props for group_by[org_unit_id]
export interface ReportData extends ReportOrgData {
  date?: string;
  values?: ReportAwsItem[] | ReportAzureItem[] | ReportGcpItem[] | ReportOcpItem[] | ReportOrgItem[];
}

export interface ReportMeta extends PagedMetaData {
  category?: string[];
  count: number;
  delta?: {
    percent: number;
    value: number;
  };
  distributed_overhead?: boolean;
  filter?: {
    [filter: string]: any;
  };
  group_by?: {
    [group: string]: string[];
  };
  limit?: number;
  offset?: number;
  order_by?: {
    [order: string]: string;
  };
  others?: number;
  total?: {
    capacity?: ReportValue;
    cost?: ReportItemValue;
    count?: ReportValue; // Workaround for https://github.com/project-koku/koku/issues/1395
    infrastructure?: ReportItemValue;
    limit?: ReportValue;
    request?: ReportValue;
    supplementary?: ReportItemValue;
    usage?: ReportValue;
  };
}

export interface Report extends PagedResponse<ReportData, ReportMeta> {}

// eslint-disable-next-line no-shadow
export const enum ReportType {
  cost = 'cost',
  cpu = 'cpu',
  database = 'database',
  instanceType = 'instance_type',
  memory = 'memory',
  network = 'network',
  org = 'org',
  storage = 'storage',
  volume = 'volume',
}

// eslint-disable-next-line no-shadow
export const enum ReportPathsType {
  aws = 'aws',
  awsOcp = 'aws_ocp',
  azure = 'azure',
  oci = 'oci',
  azureOcp = 'azure_ocp',
  gcp = 'gcp',
  gcpOcp = 'gcp_ocp',
  ibm = 'ibm',
  ocp = 'ocp',
  ocpCloud = 'ocp_cloud',
  rhel = 'rhel',
}
