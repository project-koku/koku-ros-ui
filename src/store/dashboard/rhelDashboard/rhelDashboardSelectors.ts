import type { RootState } from 'store/rootReducer';
import { getCurrency } from 'utils/localStorage';

import {
  getQueryForWidget,
  getQueryForWidgetTabs,
  rhelDashboardDefaultFilters,
  rhelDashboardStateKey,
  rhelDashboardTabFilters,
} from './rhelDashboardCommon';

export const selectRhelDashboardState = (state: RootState) => state[rhelDashboardStateKey];

export const selectWidgets = (state: RootState) => selectRhelDashboardState(state).widgets;

export const selectWidget = (state: RootState, id: number) => selectWidgets(state)[id];

export const selectCurrentWidgets = (state: RootState) => selectRhelDashboardState(state).currentWidgets;

export const selectWidgetQueries = (state: RootState, id: number) => {
  const widget = selectWidget(state, id);

  const defaultFilter = {
    ...rhelDashboardDefaultFilters,
    ...(widget.filter ? widget.filter : ({} as any)),
  };
  const tabsFilter = {
    ...rhelDashboardTabFilters,
    ...(widget.tabsFilter ? widget.tabsFilter : ({} as any)),
  };
  const props = {
    currency: getCurrency(),
  };

  return {
    previous: getQueryForWidget(
      {
        ...defaultFilter,
        time_scope_value: -2,
      },
      props
    ),
    current: getQueryForWidget(defaultFilter, props),
    forecast: getQueryForWidget({}, { limit: 31, ...props }),
    tabs: getQueryForWidgetTabs(
      widget,
      {
        ...(tabsFilter as any),
        resolution: 'monthly',
      },
      props
    ),
  };
};
