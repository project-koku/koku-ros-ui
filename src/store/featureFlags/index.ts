import * as featureFlagsActions from './featureFlagsActions';
import type { FeatureFlagsAction, FeatureFlagsState } from './featureFlagsReducer';
import { featureFlagsReducer, stateKey as featureFlagsStateKey } from './featureFlagsReducer';
import * as featureFlagsSelectors from './featureFlagsSelectors';

export type { FeatureFlagsAction, FeatureFlagsState };
export { featureFlagsActions, featureFlagsReducer, featureFlagsSelectors, featureFlagsStateKey };
