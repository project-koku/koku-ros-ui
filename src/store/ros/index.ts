import * as rosActions from './rosActions';
import { rosStateKey } from './rosCommon';
import type { CachedRos, RosAction, RosState } from './rosReducer';
import { rosReducer } from './rosReducer';
import * as rosSelectors from './rosSelectors';

export type { RosAction, CachedRos, RosState };
export { rosActions, rosReducer, rosSelectors, rosStateKey };
