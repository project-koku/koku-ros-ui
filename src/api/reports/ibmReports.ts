import axios from 'axios';

import type { Report, ReportData, ReportItem, ReportItemValue, ReportMeta, ReportValue } from './report';
import { ReportType } from './report';

export interface IbmReportItem extends ReportItem {
  account?: string;
  project?: string;
  region?: string;
  service?: string;
}

export interface GroupByAccountData extends Omit<IbmReportData, 'accounts'> {
  account: string;
}

export interface GroupByRegionData extends Omit<IbmReportData, 'regions'> {
  region: string;
}

export interface GroupByServiceData extends Omit<IbmReportData, 'services'> {
  service: string;
}

export interface GroupByProjectData extends Omit<IbmReportData, 'projects'> {
  project: string;
}

export interface IbmReportData extends ReportData {
  accounts?: GroupByAccountData[];
  projects?: GroupByProjectData[];
  regions?: GroupByRegionData[];
  services?: GroupByServiceData[];
}

export interface IbmReportMeta extends ReportMeta {
  total?: {
    cost?: ReportItemValue;
    infrastructure?: ReportItemValue;
    supplementary?: ReportItemValue;
    usage?: ReportValue;
  };
}

export interface IbmReport extends Report {
  meta: IbmReportMeta;
  data: IbmReportData[];
}

// Todo: update to use ibm when APIs are available
export const ReportTypePaths: Partial<Record<ReportType, string>> = {
  [ReportType.cost]: 'reports/gcp/costs/',
  [ReportType.database]: 'reports/gcp/costs/',
  [ReportType.network]: 'reports/gcp/costs/',
  [ReportType.storage]: 'reports/gcp/storage/',
  [ReportType.instanceType]: 'reports/gcp/instance-types/',
};

export function runReport(reportType: ReportType, query: string) {
  const path = ReportTypePaths[reportType];
  return axios.get<IbmReport>(`${path}?${query}`);
}
