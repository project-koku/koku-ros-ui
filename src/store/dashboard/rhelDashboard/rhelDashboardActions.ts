import type { ThunkAction } from 'store/common';
import { forecastActions } from 'store/forecasts';
import { reportActions } from 'store/reports';
import { createAction } from 'typesafe-actions';

import type { RhelDashboardTab } from './rhelDashboardCommon';
import { selectWidget, selectWidgetQueries } from './rhelDashboardSelectors';

export const fetchWidgetForecasts = (id: number): ThunkAction => {
  return (dispatch, getState) => {
    const state = getState();
    const widget = selectWidget(state, id);

    if (widget.forecastPathsType && widget.forecastType) {
      const { forecast } = selectWidgetQueries(state, id);
      dispatch(forecastActions.fetchForecast(widget.forecastPathsType, widget.forecastType, forecast));
    }
  };
};

export const fetchWidgetReports = (id: number): ThunkAction => {
  return (dispatch, getState) => {
    const state = getState();
    const widget = selectWidget(state, id);
    const { previous, current, tabs } = selectWidgetQueries(state, id);
    dispatch(reportActions.fetchReport(widget.reportPathsType, widget.reportType, current));
    dispatch(reportActions.fetchReport(widget.reportPathsType, widget.reportType, previous));
    if (widget.availableTabs) {
      dispatch(reportActions.fetchReport(widget.reportPathsType, widget.reportType, tabs));
    }
  };
};

export const setWidgetTab = createAction('rhelDashboard/widget/tab')<{
  id: number;
  tab: RhelDashboardTab;
}>();

export const changeWidgetTab = (id: number, tab: RhelDashboardTab): ThunkAction => {
  return dispatch => {
    dispatch(setWidgetTab({ id, tab }));
    dispatch(fetchWidgetReports(id));
  };
};
