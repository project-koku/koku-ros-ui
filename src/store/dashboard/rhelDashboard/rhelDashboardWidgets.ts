import { ForecastPathsType, ForecastType } from 'api/forecasts/forecast';
import { ReportPathsType, ReportType } from 'api/reports/report';
import messages from 'locales/messages';
import { routes } from 'routes';
import {
  ComputedReportItemType,
  ComputedReportItemValueType,
  DatumType,
} from 'routes/components/charts/common/chartDatum';
import { ComputedForecastItemType } from 'routes/components/charts/common/chartDatumForecast';
import { DashboardChartType } from 'store/dashboard/common/dashboardCommon';
import { formatCurrency, formatUnits } from 'utils/format';
import { formatPath } from 'utils/paths';

import type { RhelDashboardWidget } from './rhelDashboardCommon';
import { RhelDashboardTab } from './rhelDashboardCommon';

let currrentId = 0;
const getId = () => currrentId++;

export const costSummaryWidget: RhelDashboardWidget = {
  availableTabs: [RhelDashboardTab.projects, RhelDashboardTab.clusters],
  chartFormatter: formatCurrency,
  chartName: 'rhelCostChart',
  chartType: DashboardChartType.dailyTrend, // No longer showing infrastructure via DashboardChartType.dailyCost
  currentTab: RhelDashboardTab.projects,
  id: getId(),
  titleKey: messages.rhelDashboardCostTitle,
  forecastPathsType: ForecastPathsType.rhel,
  forecastType: ForecastType.cost,
  reportPathsType: ReportPathsType.rhel,
  reportType: ReportType.cost,
  details: {
    costKey: messages.cost,
    showHorizontal: true,
    showTooltip: true,
    viewAllPath: formatPath(routes.rhelDetails.path),
  },
  trend: {
    computedForecastItem: ComputedForecastItemType.cost,
    computedForecastInfrastructureItem: ComputedForecastItemType.infrastructure,
    computedReportItem: ComputedReportItemType.cost,
    computedReportItemValue: ComputedReportItemValueType.total,
    dailyTitleKey: messages.rhelDashboardDailyCostTitle,
    datumType: DatumType.cumulative,
    titleKey: messages.rhelDashboardCostTrendTitle,
  },
  tabsFilter: {
    limit: 3,
  },
};

export const cpuWidget: RhelDashboardWidget = {
  chartFormatter: formatUnits,
  chartName: 'rhelCpuChart',
  chartType: DashboardChartType.usage,
  id: getId(),
  titleKey: messages.rhelCpuUsageAndRequests,
  reportPathsType: ReportPathsType.rhel,
  reportType: ReportType.cpu,
  details: {
    requestKey: messages.requests,
    showUnits: true,
    showUsageFirst: true,
    usageKey: messages.usage,
  },
  trend: {
    computedReportItem: ComputedReportItemType.usage,
    computedReportItemValue: ComputedReportItemValueType.total,
    datumType: DatumType.rolling,
    titleKey: messages.rhelDailyUsageAndRequestComparison,
  },
};

export const memoryWidget: RhelDashboardWidget = {
  chartFormatter: formatUnits,
  chartName: 'rhelMemoryChart',
  chartType: DashboardChartType.usage,
  id: getId(),
  titleKey: messages.rhelMemoryUsageAndRequests,
  reportPathsType: ReportPathsType.rhel,
  reportType: ReportType.memory,
  details: {
    requestKey: messages.requests,
    showUnits: true,
    showUsageFirst: true,
    usageKey: messages.usage,
  },
  trend: {
    computedReportItem: ComputedReportItemType.usage,
    computedReportItemValue: ComputedReportItemValueType.total,
    datumType: DatumType.rolling,
    titleKey: messages.rhelDailyUsageAndRequestComparison,
  },
};

export const volumeWidget: RhelDashboardWidget = {
  chartFormatter: formatUnits,
  chartName: 'rhelVolumeChart',
  chartType: DashboardChartType.usage,
  id: getId(),
  titleKey: messages.rhelVolumeUsageAndRequests,
  reportPathsType: ReportPathsType.rhel,
  reportType: ReportType.volume,
  details: {
    requestKey: messages.requests,
    showUnits: true,
    showUsageFirst: true,
    usageKey: messages.usage,
  },
  trend: {
    computedReportItem: ComputedReportItemType.usage,
    computedReportItemValue: ComputedReportItemValueType.total,
    datumType: DatumType.rolling,
    titleKey: messages.rhelDailyUsageAndRequestComparison,
  },
};
