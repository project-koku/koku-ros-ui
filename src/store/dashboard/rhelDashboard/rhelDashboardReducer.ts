import type { ActionType } from 'typesafe-actions';
import { getType } from 'typesafe-actions';

import { setWidgetTab } from './rhelDashboardActions';
import type { RhelDashboardWidget } from './rhelDashboardCommon';
import { costSummaryWidget, cpuWidget, memoryWidget, volumeWidget } from './rhelDashboardWidgets';

export type RhelDashboardAction = ActionType<typeof setWidgetTab>;

export type RhelDashboardState = Readonly<{
  widgets: Record<number, RhelDashboardWidget>;
  currentWidgets: number[];
}>;

export const defaultState: RhelDashboardState = {
  currentWidgets: [costSummaryWidget.id, cpuWidget.id, memoryWidget.id, volumeWidget.id],
  widgets: {
    [costSummaryWidget.id]: costSummaryWidget,
    [cpuWidget.id]: cpuWidget,
    [memoryWidget.id]: memoryWidget,
    [volumeWidget.id]: volumeWidget,
  },
};

export function rhelDashboardReducer(state = defaultState, action: RhelDashboardAction): RhelDashboardState {
  switch (action.type) {
    case getType(setWidgetTab):
      return {
        ...state,
        widgets: {
          ...state.widgets,
          [action.payload.id]: {
            ...state.widgets[action.payload.id],
            currentTab: action.payload.tab,
          },
        },
      };
    default:
      return state;
  }
}
