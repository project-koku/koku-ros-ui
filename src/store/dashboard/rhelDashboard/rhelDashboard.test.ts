jest.mock('store/reports/reportActions');

import { ReportType } from 'api/reports/report';
import { ComputedReportItemType, DatumType } from 'routes/components/charts/common/chartDatum';
import { createMockStoreCreator } from 'store/mockStore';
import { reportActions } from 'store/reports';

import * as actions from './rhelDashboardActions';
import {
  getGroupByForTab,
  getQueryForWidgetTabs,
  rhelDashboardStateKey,
  RhelDashboardTab,
} from './rhelDashboardCommon';
import { rhelDashboardReducer } from './rhelDashboardReducer';
import * as selectors from './rhelDashboardSelectors';
import { costSummaryWidget, cpuWidget, memoryWidget, volumeWidget } from './rhelDashboardWidgets';

const createRhelDashboardStore = createMockStoreCreator({
  [rhelDashboardStateKey]: rhelDashboardReducer,
});

const fetchReportMock = reportActions.fetchReport as jest.Mock;

beforeEach(() => {
  fetchReportMock.mockReturnValue({ type: '@@test' });
});

test('default state', () => {
  const store = createRhelDashboardStore();
  const state = store.getState();
  expect(selectors.selectCurrentWidgets(state)).toEqual([
    costSummaryWidget.id,
    cpuWidget.id,
    memoryWidget.id,
    volumeWidget.id,
  ]);
  expect(selectors.selectWidget(state, costSummaryWidget.id)).toEqual(costSummaryWidget);
});

test('fetch widget reports', () => {
  const store = createRhelDashboardStore();
  store.dispatch(actions.fetchWidgetReports(costSummaryWidget.id));
  expect(fetchReportMock.mock.calls).toMatchSnapshot();
});

test('changeWidgetTab', () => {
  const store = createRhelDashboardStore();
  store.dispatch(actions.changeWidgetTab(costSummaryWidget.id, RhelDashboardTab.projects));
  const widget = selectors.selectWidget(store.getState(), costSummaryWidget.id);
  expect(widget.currentTab).toBe(RhelDashboardTab.projects);
  expect(fetchReportMock).toHaveBeenCalledTimes(3);
});

describe('getGroupByForTab', () => {
  test('clusters tab', () => {
    expect(getGroupByForTab(RhelDashboardTab.clusters)).toMatchSnapshot();
  });

  test('nodes tab', () => {
    expect(getGroupByForTab(RhelDashboardTab.nodes)).toMatchSnapshot();
  });

  test('projects tab', () => {
    expect(getGroupByForTab(RhelDashboardTab.projects)).toMatchSnapshot();
  });

  test('unknown tab', () => {
    expect(getGroupByForTab('unknown' as any)).toMatchSnapshot();
  });
});

test('getQueryForWidget', () => {
  const widget = {
    id: 1,
    titleKey: '',
    reportType: ReportType.cost,
    availableTabs: [RhelDashboardTab.projects],
    currentTab: RhelDashboardTab.projects,
    details: { formatOptions: {} },
    trend: {
      computedReportItem: ComputedReportItemType.cost,
      datumType: DatumType.rolling,
      formatOptions: {},
      titleKey: '',
    },
    topItems: {
      formatOptions: {},
    },
  };

  [
    [
      undefined,
      'filter[resolution]=daily&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[project]=*',
    ],
    [{}, 'group_by[project]=*'],
    [{ limit: 3 }, 'filter[limit]=3&group_by[project]=*'],
  ].forEach(value => {
    expect(getQueryForWidgetTabs(widget, value[0])).toEqual(value[1]);
  });
});
