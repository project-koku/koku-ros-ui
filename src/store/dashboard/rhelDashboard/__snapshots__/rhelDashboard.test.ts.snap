// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch widget reports 1`] = `
[
  [
    "rhel",
    "cost",
    "currency=USD&filter[resolution]=daily&filter[time_scope_units]=month&filter[time_scope_value]=-1",
  ],
  [
    "rhel",
    "cost",
    "currency=USD&filter[resolution]=daily&filter[time_scope_units]=month&filter[time_scope_value]=-2",
  ],
  [
    "rhel",
    "cost",
    "currency=USD&filter[limit]=3&filter[resolution]=monthly&filter[time_scope_units]=month&filter[time_scope_value]=-1&group_by[project]=*",
  ],
]
`;

exports[`getGroupByForTab clusters tab 1`] = `
{
  "cluster": "*",
}
`;

exports[`getGroupByForTab nodes tab 1`] = `
{
  "node": "*",
}
`;

exports[`getGroupByForTab projects tab 1`] = `
{
  "project": "*",
}
`;

exports[`getGroupByForTab unknown tab 1`] = `{}`;
